Add Unit Tests for Espresso Streamer

Ayiga opened the pull request:
<!-- Closes #<ISSUE_NUMBER> -->
<!-- These comments should help create a useful PR message, please delete any remaining comments before opening the PR. -->
<!-- If there is no issue number make sure to describe clearly *why* this PR is necessary. -->
<!-- Mention open questions, remaining TODOs, if any -->

### This PR:
<!-- Describe what this PR adds to this repo and why -->
<!-- E.g. -->
<!-- * Implements feature 1 -->
<!-- * Fixes bug 3 -->
Adds unit tests in order to effectively test the `EspressoStreamer`'s implementation with it's expected calling flows.  These tests include incrementing for every batch created, creating all of the batches, then incrementing over them, and then creating all of the batches and them being stored in a random Espresso Block, and still finding them in order.

In order to effectively test these features, this test creates a mock implementation of tracking the `L1`, `L2`, and `Espresso` state, so it can simulate and mock their responses effectively.  This required replacing pointer implementation instances of the various Espresso implementations with interface references that document only the methods being consumed to make it easier to implement.

### This PR does not:
<!-- Describe what is out of scope for this PR, if applicable. Leave this section blank if it's not applicable -->
<!-- This section helps avoid the reviewer having to needlessly point out missing parts -->
<!-- * Implement feature 3 because that feature is blocked by Issue 4 -->
<!-- * Implement xyz because that is tracked in issue #123. -->
<!-- * Address xzy for which I opened issue #456 -->
Aim to change or modify the actual logic of the EspressoStreamer, beyond replacing the pointer references to structs with interfaces, so scenarios, use cases, and edge-cases can be more easily tested.

<!-- ### Key places to review: -->
<!-- Describe key places for reviewers to pay close attention to -->
<!-- * file.rs, `add_integers` function -->
<!-- Or directly comment on those files/lines to make it easier for the reviewers -->

<!-- ### How to test this PR:  -->
<!-- Optional, uncomment the above line if this is relevant to your PR -->
<!-- If your PR is fully tested through CI there is no need to add this section -->
<!-- * E.g. `just test` -->

<!-- ### Things tested -->
<!-- Anything that was manually tested (that is not tested in CI). -->
<!-- E.g. building/running of docker containers. Changes to docker demo, ... -->
<!-- Especially mention anything untested, with reasoning and link an issue to resolve this. -->

<!-- Complete the following items before creating this PR -->
<!-- [ ] Issue linked or PR description mentions why this change is necessary. -->
<!-- [ ] PR description is clear enough for reviewers. -->
<!-- [ ] Documentation for changes (additions) has been updated (added).  -->
<!-- [ ] If this is a draft it is marked as "draft".  -->

<!-- To make changes to this template edit https://github.com/EspressoSystems/.github/blob/main/PULL_REQUEST_TEMPLATE.md -->


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1210175561822693

Ayiga committed:
Add some unit tests to Espresso Streamer

In order to increase our confidence in our implementation of the
EspressoStreamer it is beneficial to have unit tests that are able to
simulate the expected behavior, and the handling of edge cases we
anticipate (and potentially do not anticipate) occurring.

With this in mind the `EspressoStreamer` has been modified to consume
interface implementations instead of struct instances so these
components are able to be swapped out with others, specifically our
mock implementation so we can more easily and readily control the
state iteration of the system and the interactions between the L1,
L2, and Espresso.

QuentinI commented on op-node/rollup/derive/espresso_batch.go:
```diff
@@ -77,6 +77,16 @@ func BlockToEspressoBatch(rollupCfg *rollup.Config, block *types.Block) (*Espres
+// unmarshal an Espresso transaction into an EspressoBatch. The returned
+// function takes a batcherAddress as an argument to verify the signature of
+// the transaction.
+func CreateUnmarshalEspressoTransaction(batcherAddress common.Address) func(data []byte) (*EspressoBatch, error) {
```
Nit: unless this is another Go convention I'm unaware of, can we rename it to something that is `CreateNoun...` instead of `CreateVerb...`? `CreateUnmarshal...` sounds to me like it will do both of those things, i.e. create and unmarshal a transaction, instead of creating a transaction unmarshal-er.

Ayiga replied:
Sure.  I think this is fair.  I had just named it thus, as it was creating a function to perform the unmarshaling by the function it ultimately delegates to. I suppose it would be more accurate to call it `CreateUnmarshalEspressoTransactionFunction` or something, but `CreateEspressoBatchUnmarshaller` could be just as appropriate.

Ayiga replied:
Addressed in 673a7b42d77c2434b5176925fe5f1ae9651b74d1.

QuentinI commented on espresso/streamer_test.go:
```diff
@@ -0,0 +1,610 @@
+// without any errors being thrown.
+
+func TestNewEspressoStreamer(t *testing.T) {
+	defer func() {
```
Is this somehow different from just writing `_, err := NewEspressoStreamer()` and checking the error?

Ayiga replied:
It is in a way.  Generally speaking, the call to `NewEspressoStreamer` doesn't have an explicit error case where it able to return an error.  In treating `NewEspressoStreamer` as a black box, we know that it cannot return an error explicitly due to the signature, but there are no guarantees that its call won't `panic`.

The deferred call to `recover` will catch all panics that occur, and allow us to report them. Now, in inspecting the code currently, this doesn't seem like it could ever happen.  But, having this check here, prevents us from introducing such a "feature" in the future.

Ayiga replied:
Though, to be fair.  I imagine that `panic` would generate an automatic fail anyway, so this is likely unnecessary entirely.

Ayiga replied:
Addressed in d60fdec94fd542878bb99c87e1b962e2e3ab95b0.

QuentinI commented on espresso/streamer_test.go:
```diff
@@ -0,0 +1,610 @@
+// AdvanceSafeL1ByNBlocks advances the SafeL1 block reference by n blocks.
+func (m *MockStreamerSource) AdvanceSafeL1ByNBlocks(n int) {
+	for i := 0; i < n; i++ {
+		m.SafeL1 = createL1BlockRef(m.SafeL1.Number + 1)
```
Nit: why not just `m.SafeL1 = createL1BlockRef(m.SafeL1.Number + n)`? This looks a bit confusing to me

Ayiga replied:
This is fair.  The only reason to play through the iteration would be if the fields that are populated aren't able to be generated deterministically, **OR** if they have side-effects.  In this case, neither apply, so I can just do as you suggest instead.

Ayiga replied:
Addressed in 14b47e608c103b37670e764b3cd2709f868392a8.

QuentinI commented on espresso/streamer_test.go:
```diff
@@ -0,0 +1,610 @@
+// AdvanceL2ByNBlocks advances the SafeL2 block reference by n blocks.
+func (m *MockStreamerSource) AdvanceL2ByNBlocks(n int) {
+	for i := 0; i < n; i++ {
+		m.SafeL2 = createL2BlockRef(m.SafeL2.Number+1, m.SafeL1)
```
Same question here

Ayiga replied:
Addressed in 14b47e608c103b37670e764b3cd2709f868392a8.

QuentinI commented on espresso/streamer_test.go:
```diff
@@ -0,0 +1,610 @@
+	// At the moment the Streamer utilizes the SyncStatus in order to update
+	// it's local state.  But, in general the Streamer doesn't consume all
+	// of the fields provided within the SyncStatus.  At the moment it only
+	// cares about SafeL2, FinalizedL1, SafeL1. So this is what we will track
```
A bit of a digression, but do we actually care for SafeL1? I've only found a place where we log it for some reason, but I don't think we use it for any logic

Ayiga replied:
I was keeping it, specifically because of the log. But I think we got rid of that log, so we can likely do away with it.

Ayiga replied:
Addressed in 676a65afbd29676ef9acfa7346d3cc0e7cf8111a.

QuentinI commented on espresso/streamer_test.go:
```diff
@@ -0,0 +1,610 @@
+	signer := geth_types.NewLondonSigner(chainID)
+	baseFee := big.NewInt(rng.Int63n(300_000_000_000))
+	txsEncoded := make([]hexutil.Bytes, 0, txCount)
+	// force each tx to have equal chainID
```
Is this comment in the right place?

Ayiga replied:
No, it's a failure to clean up some previous code.

Ayiga replied:
Addressed in 676a65afbd29676ef9acfa7346d3cc0e7cf8111a.

QuentinI reviewed (approved):
This is great! Left some nitpicks

Ayiga committed:
Remove unnecessary for loops in mock source

In review it was pointed out by @QuentinI that incrementing through a for
loop to change the position by one incrementally was unnecessary, and
potentially confusing:
https://github.com/EspressoSystems/optimism-espresso-integration/pull/125#discussion_r2075881991
https://github.com/EspressoSystems/optimism-espresso-integration/pull/125#discussion_r2075743297

The only reason to have an incremental approach like this would be due
some non deterministic reference creation, or if there are side-effects
being triggered by the advancing.  This is not currently the case for
either of these calls, and this makes the for loop entirely unnecessary
and potentially confusing.

Ayiga committed:
Rename function to create Unmarshal function

@QuentinI has pointed out that the name of the function
`CreateUnmarshalEspressoTransaction` fails to effectively convey what
is actually happening by this function in an effective manner, as it is
named as if it it may be performing multiple actions:

https://github.com/EspressoSystems/optimism-espresso-integration/pull/125#discussion_r2075721762

This change renames the function in question in hopes to clear up the
ambiguity .

Ayiga committed:
Remove SafeL1 from MockStreamerSource

As pointed out by @QuentinI, the reference to the `SafeL1` is not
relevant.  It was originally added as the `Refresh` call made a reference
to it for logging originally.  This reference and log were removed thereby
removing the need for the `SafeL1` tracking at all.

https://github.com/EspressoSystems/optimism-espresso-integration/pull/125#discussion_r2075750502

Replace all former references of the `SafeL1` with teh `FinalizedL1`
within streamer_test.go.

Ayiga commented:
I would love for some suggestions about specific edge cases / scenarios we should add.

Ayiga committed:
Remove unnecessary defer recover in test

Based on the discussion in the Pull Request, this panic is ultimately not needed as any
panic would automatically be caught by the test and reported accordingly.

https://github.com/EspressoSystems/optimism-espresso-integration/pull/125#discussion_r2075729141

Remove the panic, and the additional comment reiterating what was being stated on the
test itself.

Ayiga committed:
Bump up the number of iterations to 1000

philippecamacho commented on espresso/streamer_test.go:
```diff
@@ -0,0 +1,589 @@
+
+// L1 Client methods
+
+func (m *MockStreamerSource) HeaderHashByNumber(ctx context.Context, number *big.Int) (common.Hash, error) {
```
Maybe rename the function to `L1HeaderHashByNumber`?

Ayiga replied:
In this case, sadly not.  This method exists to adhere to the interface that the `EspressoStreamer` is consuming from.  So changing the name would mean it no longer adheres to the interface.

We could change the name, but it's something @dailinsubjam named, and it would require us to modify it in the three places that it is referenced.  Though, this may be worth while since it's all code we introduced, and it helps to clarify the specific functionality.

philippecamacho replied:
Sure, let us keep it as is.

philippecamacho commented on espresso/streamer_test.go:
```diff
@@ -0,0 +1,589 @@
+
+// NoOpLogger is a no-op implementation of the log.Logger interface.
+// It is used to pass a non-nil logger to the EspressoStreamer without
+// producing any output.
```
What is the benefit of this NoOpLogger?

Ayiga replied:
No direct benefit, other than preventing the spam of logging.  I primarily did it to be able to pass a Logger that adheres to the correct interface, without having to jump through the hoops of figuring out how to create an actual logger.

(We don't do `nil` checks when logging, so this **MUST** not be `nil`)

philippecamacho commented on espresso/streamer_test.go:
```diff
@@ -0,0 +1,589 @@
+			}
+
+			state.AdvanceSafeL2()
+		}
```
Any reason why we don't advance the finalized L1 head? 
```
state.AdvanceFinalizedL1()
```

Ayiga replied:
No particular reason, other than it seems like there isn't any direct need to in order to verify behavior.  We can add another test to advance the l1 to see how it affects things as well.  However, in order for it to have any affect on the `EspressoStreamer`, we would need to be calling `.Refresh` within the loop some where.

philippecamacho commented:
This is absolutely great!
Do you think it is possible to easily mock the `CheckBatch` function at some point in order to check the logic of updating the batch buffer and remaining list? This would be done definitely in another PR.

philippecamacho reviewed (approved):
LGTM
